dialogue-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientBlockingChannel.java

```diff
@@ -0,0 +1,199 @@
+        public int code() {
+            return response.getStatusLine().getStatusCode();
+        }
+
+        @Override
+        public Map<String, List<String>> headers() {
+            if (headers == null) {
+                Map<String, List<String>> tmpHeaders = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+                for (Header header : response.getAllHeaders()) {
+                    String value = header.getValue();
+                    if (value != null) {
+                        tmpHeaders
+                                .computeIfAbsent(header.getName(), _name -> new ArrayList<>(1))
+                                .add(header.getValue());
+                    }
+                }
+                headers = tmpHeaders;
+            }
+            return headers;
+        }
+
+        @Override
+        public Optional<String> getFirstHeader(String header) {
+            return Optional.ofNullable(response.getFirstHeader(header)).map(Header::getValue);
+        }
+
+        @Override
+        public void close() {
+            try {
+                response.close();
+            } catch (IOException | RuntimeException e) {
+                log.warn("Failed to close response", e);
+            }
+        }
+
+        @Override
+        public String toString() {
+            return "HttpClientResponse{response=" + response + '}';
+        }
+    }
+
+    private static final class RequestBodyEntity implements HttpEntity {
+
+        private final RequestBody requestBody;
+        private final Header contentType;
+
+        RequestBodyEntity(RequestBody requestBody) {
+            this.requestBody = requestBody;
+            this.contentType = new BasicHeader(Headers.CONTENT_TYPE, requestBody.contentType());
+        }
+
+        @Override
+        public boolean isRepeatable() {
+            return false;
+        }
+
+        @Override
+        public boolean isChunked() {
+            return true;
+        }
+
+        @Override
+        public long getContentLength() {
+            // unknown
+            return -1;
+        }
+
+        @Override
+        public Header getContentType() {
+            return contentType;
+        }
+
+        @Override
+        public Header getContentEncoding() {
+            return null;
+        }
+
+        @Override
+        public InputStream getContent() throws UnsupportedOperationException {
+            throw new UnsupportedOperationException("getContent is not supported, writeTo should be used");
+        }
+
+        @Override
+        public void writeTo(OutputStream outStream) throws IOException {
+            requestBody.writeTo(outStream);
+        }
+
+        @Override
+        public boolean isStreaming() {
+            // Applies to responses.
+            return false;
+        }
+
+        @Override
+        public void consumeContent() {}
+
+        @Override
+        public String toString() {
+            return "RequestBodyEntity{requestBody=" + requestBody + '}';
```

Comment:
I'd love to get your immutable RequestBody PR merged, because I think as we have no control over the requestBody implementation, there's currently a danger that this toString value could include sensitive headers (incl. bearertokens).

Reply:
We did! This is the body object (tuple of consumer<OutputStream> and contentType)

Reply:
ah cool sorry misread as request 🤦‍♂